crates/red_knot_python_semantic/src/semantic_index/use_def/symbol_state.rs

```diff
@@ -0,0 +1,362 @@
+//! Track visible definitions of a symbol, and applicable constraints per definition.
+//!
+//! These data structures operate entirely on scope-local newtype-indices for definitions and
+//! constraints, referring to their location in the `all_definitions` and `all_constraints`
+//! indexvecs in [`super::UseDefMapBuilder`].
+//!
+//! We need to track arbitrary associations between definitions and constraints, not just a single
+//! set of currently dominating constraints (where "dominating" means "control flow must have
+//! passed through it to reach this point"), because we can have dominating constraints that apply
+//! to some definitions but not others, as in this code:
+//!
+//! ```python
+//! x = 1 if flag else None
+//! if x is not None:
+//!     if flag2:
+//!         x = 2 if flag else None
+//!     x
+//! ```
+//!
+//! From the point of view of the final use of `x`, the `x is not None` constraint is "currently
+//! visible" (in the sense that control flow must have passed through it to reach this point), but
+//! it applies only to the first definition of `x`, not the second, so `None` is a possible value
+//! for `x`.
+//!
+//! And we can't just track an index into a list of dominating constraints for each definition,
+//! either, because we can have definitions which are still visible, but subject to constraints
+//! that are no longer dominating, as in this code:
+//!
+//! ```python
+//! x = 0
+//! if flag1:
+//!     x = 1 if flag2 else None
+//!     assert x is not None
+//! x
+//! ```
+//!
+//! From the point of view of the final use of `x`, the `x is not None` constraint no longer
+//! dominates, but it does dominate the `x = 1 if flag2 else None` definition, so we have to keep
+//! track of that.
+//!
+//! The data structures used here ([`BitSet`] and [`BitSetArray`]) optimize for keeping all data
+//! inline (avoiding lots of scattered allocations) in small-to-medium cases, and falling back to
+//! heap allocation to be able to scale to arbitrary numbers of definitions and constraints when
+//! needed.
+use super::bitset::{BitSet, BitSetArray, BitSetArrayIterator, BitSetIterator};
+use ruff_index::newtype_index;
+
+/// A newtype-index for a definition in a particular scope.
+#[newtype_index]
+pub(super) struct ScopedDefinitionId;
+
+/// A newtype-index for a constraint expression in a particular scope.
+#[newtype_index]
+pub(super) struct ScopedConstraintId;
+
+/// Can reference this * 64 total definitions inline; more will fall back to the heap.
+const INLINE_DEFINITION_BLOCKS: usize = 1;
```

Comment:
Could you let me know how you determined this number? We could probably go higher than this because the size of a `BTreeSet` is 24 bytes. This means we should use at least two blocks to make the best use of the stack memory. Whether we want to use 3 depends if Rust is able to use a niche to represent the `BitSet` enum as 24 bytes if the array is small enough. If not, then we should use 3 blocks at least.

Reply:
I think I initially had this set as 4, and found that 1 performed better, and 1-3 didn't show any meaningful difference. It's a good point though that it should be at least as big as the `BTreeSet` (though this factor may disappear if I go with your growable-bitset suggestion.)

Reply:
I think currently, given the factor I mentioned in our 1:1 (that the performance bottleneck in this PR is cloning the entire IndexVec of SymbolStates at every branch point), the dominating factor in performance is to reduce the size of that `IndexVec`, which means using as little memory here as possible seems to be more important than keeping the data always inline (though probably only to a point; I suspect it would be a regression if we always had to allocate everything as vectors-of-vectors.)

Reply:
A `BitSet<1>` is 32 bytes, suggesting Rust isn't able to use a niche to differentiate the enum. And indeed, BitSets of size 1, 2, and 3 are all 32 bytes; 4 increases to 40 bytes.

Empirically I'm not able to detect any performance difference between 1, 2, 3, or 4. I'll leave it at 3 for now; 3 * 64 is quite a lot of definitions in one scope.

Reply:
I'll revisit this after I implement the growable-bitset version.